### PR TITLE
[CI] Provide XML configuration for QEMU

### PR DESCRIPTION
**Pull request by KFilipek**
#### Description
This pull request introduces a new feature to provide XML-based configuration.
The `qemu_config.py` script takes provided XML config and prints QEMU arguments.

### Todo
- [x] Write a script to provide args
- [x] Add default config
- [x] Run successfully QEMU and execute tests as previously
- [x] Replace previous workflow
- [x] Cleanup and undraft

### Checklist

- [ ] Code compiles without errors locally
- [ ] All tests pass locally
- [ ] CI workflows execute properly

- [ ] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [ ] Added/extended example(s) to cover this functionality
- [ ] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [ ] Logger (with debug/info/... messages) is used

### CODE REVIEW

File: scripts/qemu/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+psutil
```
Comment (bratpiorka) [inline scripts/qemu/requirements.txt:1]: please add an empty line at the end of this file

File: scripts/qemu/configs/default.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        </distances>
+      </cell>
+      <cell id='2' memory='1200' unit='MiB'>
+        <distances>
+          <sibling id='0' value='17'/>
+          <sibling id='1' value='28'/>
+          <sibling id='2' value='10'/>
+        </distances>
+      </cell>
+        <!--
```
Comment (bratpiorka) [inline scripts/qemu/configs/default.xml:35]: please remove this comment

File: scripts/qemu/requirements.txt (hunk 1 above)
Comment (KFilipek) [reply scripts/qemu/requirements.txt:1]: Done

File: scripts/qemu/configs/default.xml (hunk 2 above)
Comment (KFilipek) [reply scripts/qemu/configs/default.xml:35]: Done

**Review by bratpiorka — APPROVED** [review]

**Review by kswiecicki — APPROVED** [review]

File: scripts/qemu/configs/default.xml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      <interconnects>
+        <latency initiator='0' target='0' type='access' value='10'/>
+        <bandwidth initiator='0' target='0' type='access' value='10485760' unit='B'/>
+        
+        <latency initiator='0' target='1' type='access' value='20'/>
+        <bandwidth initiator='0' target='1' type='access' value='5242880' unit='B'/>
+        
+        <latency initiator='0' target='2' type='access' value='16'/>
+        <bandwidth initiator='0' target='2' type='access' value='1048576' unit='B'/>
+
```
Comment (lukaszstolarczuk) [inline scripts/qemu/configs/default.xml:44]: I think we're missing
```
        -numa hmat-lb,initiator=1,target=0,hierarchy=memory,data-type=access-latency,latency=20 \
        -numa hmat-lb,initiator=1,target=0,hierarchy=memory,data-type=access-bandwidth,bandwidth=5242880
```

Comment (KFilipek) [reply scripts/qemu/configs/default.xml:44]: It's intentional, you cannot add those lines :)

File: scripts/qemu/configs/default.xml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<!-- SPDX-License-Identifier: BSD-2-Clause
+# Copyright (C) 2024 Intel Corporation.
+-->
+<domain type='qemu'>
+  <name>CascadeLake2Var1</name>
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/configs/default.xml:5]: Copied from Memkind, fixed

File: scripts/qemu/qemu_config.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
+import re
+import subprocess  # nosec
+import sys
+import collections
+import os
+import psutil  # pip install psutil
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:6]: Done

File: scripts/qemu/qemu_config.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+verbose_mode = False
+
+
+def enable_verbose():
+    """
+    Parse command line arguments
+    """
+    # print(os.getenv('ENABLE_VERBOSE', 'False'))
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:23]: Done

File: scripts/qemu/qemu_config.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
+import re
+import subprocess  # nosec
+import sys
+import collections
+import os
+import psutil  # pip install psutil
+import shutil
+
+# First of all: pip install -r requirements.txt
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:9]: I rewrote this line to be more clear

File: scripts/qemu/qemu_config.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    global verbose_mode
+    verbose_mode = os.getenv("ENABLE_VERBOSE", False)
+
+
+def parse_topology_xml(tpg_file_name: str) -> TopologyCfg:
+    """
+    Parse topology xml file
+    """
+    try:
+        virsh_path = shutil.which("virsh")
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:33]: I've added line above to comment section

File: scripts/qemu/qemu_config.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        if verbose_mode != False:
+            print(f"Name: {tpg_cfg['cpu_model']}")
+            print(f"HMAT: {tpg_cfg['hmat']}")
+            print(f"CPU_MODEL: {tpg_cfg['cpu_model']}")
+            print(f"CPU_OPTIONS: {tpg_cfg['cpu_options']}")
+            print(f"MEM_OPTIONS: {tpg_cfg['mem_options']}")
+
+        tpg = TopologyCfg(**tpg_cfg)
+    except FileNotFoundError:
+        sys.exit(f"\n Provided file is missing or missing virsh.")
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:67]: I just copied it from Memkind script.

File: scripts/qemu/qemu_config.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            raise MemoryHostException(mem.total, mem_needed, tpg.name)
+        return f"-m {mem_needed/1024/1024}M"
+    else:
+        return "-m 2G"
+
+
+if __name__ == "__main__":
+    enable_verbose()
+
+    # Parse arg named: tpg_file_name
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:103]: Copilot prompt, sorry

File: scripts/qemu/qemu_config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply scripts/qemu/qemu_config.py]: thx!

File: scripts/qemu/qemu_config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply scripts/qemu/qemu_config.py:68]: I guess there was an issue there, as well. Nonetheless, it dosn't seem to be caught here.

**Comment by lukaszstolarczuk** [comment]
pls rebase, I think GHA made some bad merge/rebase and that's why QEMU failed

File: scripts/qemu/qemu_config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KFilipek) [reply scripts/qemu/qemu_config.py:68]: I see, I moved it below, as a generic exception type.

File: .github/workflows/qemu.yml [hunk 11]
```diff
@@ -65,6 +70,9 @@ jobs:
       run: wget https://cloud-images.ubuntu.com/releases/lunar/release/ubuntu-23.04-server-cloudimg-amd64.img
     - name: Resize image
       run: qemu-img resize ./ubuntu-23.04-server-cloudimg-amd64.img +4G
+    - name: Print qemu args
+      run: |
+        echo `python3 scripts/qemu/qemu_config.py scripts/qemu/configs/default.xml | sed s/''\''/'/g`
     - name: Run qemu
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply .github/workflows/qemu.yml:76]: bump - perhaps we shouldn't define the model of the CPU - previously we didn't set a model or its topology

[the remark this answers is not in the document]
Comment (KFilipek) [reply .github/workflows/qemu.yml:76]: You were right, done

File: scripts/qemu/configs/default.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
+<!-- SPDX-License-Identifier: BSD-2-Clause
+# Copyright (C) 2024 Intel Corporation.
+-->
+<domain type='qemu'>
+  <name>CascadeLake</name>
```
Comment (lukaszstolarczuk) [inline scripts/qemu/configs/default.xml:5]: nit: when you'll be adding more configs, pls re-name this one to e.g. "Default"

**Review by lukaszstolarczuk — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

File: scripts/qemu/configs/default.xml (hunk 12 above)
Comment (KFilipek) [reply scripts/qemu/configs/default.xml:5]: Okay, thanks!

**Pull request merged**